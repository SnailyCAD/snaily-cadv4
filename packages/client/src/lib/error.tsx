import { X } from "react-bootstrap-icons";
import toast, { type ToastOptions, ErrorIcon } from "react-hot-toast";

interface Options extends ToastOptions {
  message: string;
  title?: string;
}

export function toastError(options: Options) {
  const { title, message, ...rest } = options;

  return toast.custom((t) => {
    return (
      <div
        className={`${
          t.visible ? "animate-enter" : "animate-leave"
        } max-w-md w-full bg-white dark:bg-dark-gray shadow-lg rounded-lg pointer-events-auto flex ring-1 ring-black ring-opacity-5`}
      >
        <div className="flex-1 w-0 p-4">
          <div className="flex items-start">
            <div className="flex-shrink-0 pt-0.5">{rest.icon !== null ? <ErrorIcon /> : null}</div>
            <div className="ml-3 flex-1 -mt-1">
              {title ? (
                <p className="capitalize text-[1.05em] mb-1 font-semibold text-gray-900 dark:text-gray-100">
                  {title}
                </p>
              ) : null}
              <p className="text-lg text-neutral-500 dark:text-gray-300">{message}</p>
            </div>
          </div>
        </div>
        <div className="flex items-start justify-start h-full">
          <button
            aria-label="Close message"
            onClick={() => toast.dismiss(t.id)}
            className="w-full border border-transparent rounded-none rounded-r-lg p-4 flex items-center justify-center text-sm font-medium"
          >
            <X
              width={25}
              height={25}
              className="fill-current dark:text-gray-300 dark:hover:text-gray-100"
            />
          </button>
        </div>
      </div>
    );
  }, rest);
}
