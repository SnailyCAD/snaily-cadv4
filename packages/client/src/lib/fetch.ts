import axios, { type Method, type AxiosRequestConfig, type AxiosResponse } from "axios";
import type { IncomingMessage } from "connect";
import nookies from "nookies";
import type { NextApiRequestCookies } from "next/dist/server/api-utils";
import { IFRAME_COOKIE_NAME } from "../pages/api/token";

export type RequestData = Record<string, unknown>;

interface Options extends Omit<AxiosRequestConfig, "headers"> {
  headers?: any;
  req?: IncomingMessage & { cookies?: NextApiRequestCookies };
  method?: Method | string;
  data?: RequestData;
  isSsr?: boolean;
}

export async function handleRequest<T = any>(
  path: string,
  options?: Options,
): Promise<AxiosResponse<T, T>> {
  const { req, method, data } = options ?? {};

  const apiUrl = findAPIUrl();
  const location = typeof window !== "undefined" ? window.location : null;
  const isDispatchUrl = (location?.pathname ?? req?.url) === "/dispatch";
  let parsedCookie = req?.headers.cookie;
  const cookies = nookies.get({ req });

  if (process.env.IFRAME_SUPPORT_ENABLED === "true" && !parsedCookie) {
    parsedCookie = cookies[IFRAME_COOKIE_NAME]!;
  }

  try {
    const res = await axios({
      url: `${apiUrl}${path}`,
      method,
      data: data ?? undefined,
      withCredentials: true,
      params: options?.params,
      headers: {
        Session: parsedCookie ?? "",
        "Content-Type": "application/json",
        "is-from-dispatch": String(isDispatchUrl),
      },
    });

    return makeReturn(res);
  } catch (e) {
    return makeReturn(e);
  }
}

export function findAPIUrl() {
  const envUrl = process.env.NEXT_PUBLIC_PROD_ORIGIN ?? "http://localhost:8080/v1";

  if (process.env.NODE_ENV === "development") {
    return "http://localhost:8080/v1";
  }

  return envUrl;
}

function makeReturn<T>(v: any): Omit<AxiosResponse<T>, "request"> {
  delete v.request;

  return v;
}
