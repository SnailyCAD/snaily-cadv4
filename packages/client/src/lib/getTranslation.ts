import { readFile } from "node:fs/promises";
import path from "node:path";
import { getNextI18nConfig } from "./i18n/getNextI18nConfig";

const cwd = process.cwd();

export async function getTranslations(types: string[], locale = "en") {
  const typesWithCommon = [...new Set(["common", "auth", ...types])];
  const paths = typesWithCommon.map((type) => path.join(cwd, `locales/${locale}/${type}.json`));
  const i18n = await getNextI18nConfig();

  if (!i18n.locales.includes(locale)) {
    locale = i18n.defaultLocale;
  }

  let data = {};

  await Promise.all(
    paths.map(async (path) => {
      const json = JSON.parse(await readFile(path, "utf8"));
      data = { ...data, ...json };
    }),
  );

  return data;
}
