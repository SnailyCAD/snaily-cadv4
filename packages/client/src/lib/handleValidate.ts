export const handleValidate =
  <Values = any>(schema: any) =>
  (values: Values) => {
    const errors: Record<string, string> = {};

    try {
      schema.parse(values);
    } catch (e: any) {
      for (const error of e.errors) {
        const [path] = error.path;
        errors[path as string] = error.message;
      }
    }

    return errors;
  };
