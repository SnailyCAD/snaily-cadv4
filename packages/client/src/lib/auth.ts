import { handleRequest } from "lib/fetch";
import type { IncomingMessage } from "node:http";
import type { GetUserData, PostUserLogoutData } from "@snailycad/types/api";

export async function getSessionUser(req?: IncomingMessage): Promise<GetUserData | null> {
  try {
    const response = await handleRequest<GetUserData | null>("/user", {
      req,
      method: "POST",
    });

    if (response.data) {
      return response.data ?? null;
    }

    return null;
  } catch {
    return null;
  }
}

export async function logout() {
  try {
    await handleRequest<PostUserLogoutData>("/user/logout", { method: "POST" });

    return true;
  } catch {
    return false;
  }
}
