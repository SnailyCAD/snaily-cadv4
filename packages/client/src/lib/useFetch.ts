import * as React from "react";
import type { AxiosRequestConfig, AxiosError } from "axios";
import { handleRequest } from "./fetch";
import { type TranslationValues, useTranslations } from "use-intl";
import Common from "../../locales/en/common.json";
import type { FormikHelpers } from "formik";
import { toastMessage } from "./toastMessage";
import { useModal } from "state/modalState";
import { ModalIds } from "types/ModalIds";

interface UseFetchOptions {
  overwriteState: State | null;
}

type NullableAbortController = AbortController | null;
type State = "loading" | "error";
export type ErrorMessage = keyof typeof import("../../locales/en/common.json")["Errors"];

type Options<Helpers extends object = object> = AxiosRequestConfig & {
  path: string;
  noToast?: boolean | ErrorMessage | (string & {});
  helpers?: FormikHelpers<Helpers>;
};

interface ErrorObj {
  message: ErrorMessage;
  data: TranslationValues;
}

interface ErrorResponseData {
  name: string;
  message: string;
  status: number;
  errors: Record<string, ErrorMessage | ErrorObj>[];
  stack: string;
}

interface Return<Data> {
  json: Data;
  error: null | ErrorMessage | (string & {});
}

export default function useFetch({ overwriteState }: UseFetchOptions = { overwriteState: null }) {
  const [state, setState] = React.useState<State | null>(null);
  const { openModal } = useModal();

  const t = useTranslations("Errors");
  const abortControllerRef = React.useRef<NullableAbortController>(null);

  React.useEffect(() => {
    setState(overwriteState);
  }, [overwriteState]);

  async function execute<Data, Helpers extends object = object>(
    options: Options<Helpers>,
  ): Promise<Return<Data>> {
    setState("loading");
    abortControllerRef.current = new AbortController();
    const { path, ...restOptions } = options;

    const mergedOptions = { ...restOptions, signal: abortControllerRef.current.signal };

    const response = await handleRequest(path, { ...mergedOptions }).catch((e) => {
      setState("error");
      return e;
    });

    const error = isAxiosError<ErrorResponseData | null>(response) ? parseError(response) : null;

    if (error) {
      const errors = parseErrors(response);
      const errorTitle = parseErrorTitle(response);

      const hasKey = isErrorKey(error);
      const key = hasKey ? error : "unknown";
      const errorObj = getErrorObj(response);

      console.error(JSON.stringify({ DEBUG: errorObj }, null, 2));

      if (response?.response?.status === 401) {
        openModal(ModalIds.ReauthorizeSession);
      }

      let hasAddedError = false as boolean; // as boolean because eslint gets upset otherwise.
      for (const error of errors) {
        Object.entries(error).map(([key, value]) => {
          const translationOptions = typeof value === "string" ? undefined : value.data;
          const translationKey = typeof value === "string" ? value : value.message;

          const message = isErrorKey(translationKey)
            ? t(translationKey, translationOptions)
            : translationKey;

          if (message && restOptions.helpers) {
            restOptions.helpers.setFieldError(key, message);
            hasAddedError = true;
          }
        });
      }

      if (
        typeof restOptions.noToast === "string" &&
        restOptions.noToast !== error &&
        !hasAddedError
      ) {
        toastMessage({ message: t(key), title: errorTitle });
      } else if (!restOptions.noToast && !hasAddedError) {
        toastMessage({ message: t(key), title: errorTitle });
      }

      setState("error");

      return {
        json: {} as Data,
        error: isAxiosError<ErrorResponseData | null>(response) ? parseError(response) : null,
      };
    }

    setState(null);

    return {
      json: response?.data ?? {},
      error: null,
    };
  }

  React.useEffect(() => {
    return () => {
      if (abortControllerRef.current) {
        abortControllerRef.current.abort();
      }
    };
  }, []);

  return { execute, state };
}

function parseError(
  error: AxiosError<ErrorResponseData | null>,
): ErrorMessage | "unknown" | (string & {}) {
  const message = error.response?.data?.message ?? error.message;
  const name = error.name;

  if (name && !message && ["NOT_FOUND", "Error"].includes(name)) {
    return name;
  }

  return message || "unknown";
}

function parseErrors(error: AxiosError<ErrorResponseData | null>) {
  return error.response?.data?.errors ?? [];
}

function parseErrorTitle(error: AxiosError<ErrorResponseData | null>) {
  const name = (error.response?.data?.name ?? error.message) as string | undefined;
  if (!name) return;

  return name.toLowerCase().replace(/_/g, " ");
}

function isAxiosError<T>(error: unknown): error is AxiosError<T, T> {
  if (!error) return false;
  return error instanceof Error || (typeof error === "object" && "response" in error);
}

function isErrorKey(key: string | ErrorObj): key is ErrorMessage {
  if (typeof key !== "string") return false;
  return Object.keys(Common.Errors).includes(key);
}

function getErrorObj(error: unknown) {
  let errorObj = {};

  if (isAxiosError(error)) {
    errorObj = {
      message: error.message,
      status: error.response?.status,
      response: error.response,
      method: error.config.method,
      data: error.config.data,
      url: error.config.url,
    };
  }

  return errorObj;
}
