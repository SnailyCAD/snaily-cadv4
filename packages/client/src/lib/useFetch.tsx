/* eslint-disable @typescript-eslint/ban-types */
import * as React from "react";
import { AxiosRequestConfig, AxiosError } from "axios";
import { handleRequest } from "./fetch";
import toast from "react-hot-toast";
import { useTranslations } from "use-intl";
import Common from "../../locales/en/common.json";

interface UseFetchOptions {
  overwriteState: State | null;
}

type NullableAbortController = AbortController | null;
type State = "loading" | "error";
export type ErrorMessage = keyof typeof import("../../locales/en/common.json")["Errors"];

type Options = AxiosRequestConfig & { noToast?: boolean };
type Return<Data> = {
  json: Data;
  error: null | ErrorMessage | (string & {});
};

export default function useFetch({ overwriteState }: UseFetchOptions = { overwriteState: null }) {
  const [state, setState] = React.useState<State | null>(null);

  const t = useTranslations("Errors");
  const abortControllerRef = React.useRef<NullableAbortController>(null);

  React.useEffect(() => {
    setState(overwriteState);
  }, [overwriteState]);

  async function execute<Data = any>(path: string, options: Options): Promise<Return<Data>> {
    setState("loading");
    abortControllerRef.current = new AbortController();

    const mergedOptions = { ...options, signal: abortControllerRef.current.signal };

    const response = await handleRequest(path, { ...mergedOptions }).catch((e) => {
      setState("error");
      return e;
    });

    const error = response instanceof Error ? parseError(response as AxiosError) : null;

    if (error) {
      const hasKey = Object.keys(Common.Errors).some((e) => e === error);
      const key = hasKey ? error : "unknown";
      const errorObj = getErrorObj(response);
      console.error({ DEBUG: JSON.stringify(errorObj, null, 2) });

      if (!options.noToast) {
        toast.error(t(key));
      }

      setState("error");

      return {
        json: {} as Data,
        error: response instanceof Error ? parseError(response as AxiosError) : null,
      };
    }

    setState(null);

    return {
      json: response?.data ?? {},
      error: null,
    };
  }

  React.useEffect(() => {
    return () => {
      if (abortControllerRef.current) {
        abortControllerRef.current.abort();
      }
    };
  }, []);

  return { execute, state };
}

function parseError(error: AxiosError): ErrorMessage | "unknown" {
  return error.response?.data?.message ?? "unknown";
}

function getErrorObj(error: unknown) {
  let errorObj = {};

  if (error instanceof Error) {
    const err = error as AxiosError;

    errorObj = {
      message: err.message,
      status: err.response?.status,
      response: err.response,
      method: err.config.method,
      data: err.config.data,
      url: err.config.url,
    };
  }

  return errorObj;
}
