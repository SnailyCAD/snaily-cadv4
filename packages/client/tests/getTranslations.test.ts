import { getTranslations } from "../src/lib/getTranslation";
import { expect, test } from "vitest";

test("Should return translations for calls & admin in EN", async () => {
  expect(await getTranslations(["calls", "admin"], "en")).toMatchSnapshot();
});

test("Should return translations for citizen in RU", async () => {
  expect(await getTranslations(["citizen"], "ru")).toMatchSnapshot();
});
