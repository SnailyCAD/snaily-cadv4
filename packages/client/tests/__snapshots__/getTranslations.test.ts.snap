// Vitest Snapshot v1

exports[`Should return translations for calls & admin in EN 1`] = `
{
  "Calls": {
    "911CallCreated": "Successfully created the 911 call.",
    "active911Calls": "Active 911 Calls",
    "addEvent": "Add Event",
    "alert_deleteCallEvent": "Are you sure you want to delete this event? This action cannot be undone.",
    "alert_end911Call": "Are you sure you want to end this call?",
    "alert_endTowCall": "Are you sure you want to end this call?",
    "alert_purgeSelectedCalls": "Are you sure you want to purge {length} item(s)? This will delete {length} item(s) from the database. This action cannot be undone.",
    "assignToCall": "Assign to call",
    "assignedUnit": "Assigned Unit",
    "assignedUnits": "Assigned Units",
    "callEvents": "Call Events",
    "callFilters": "Call Filters",
    "callTow": "Call Tow",
    "caller": "Caller",
    "citizen": "Citizen",
    "create911Call": "Create 911 call",
    "createTaxiCall": "Create taxi call",
    "createTowCall": "Create tow call",
    "deleteCallEvent": "Delete Event",
    "deliveryAddress": "Delivery Address",
    "departments": "Departments",
    "divisions": "Divisions",
    "editTaxiCall": "Edit taxi call",
    "editTowCall": "Edit tow call",
    "end911Call": "End 911 Call",
    "endCall": "End Call",
    "location": "Location",
    "manage911Call": "Manage 911 call",
    "no911Calls": "There are no 911 calls yet.",
    "noEvents": "This call does not have any events",
    "noTaxiCalls": "There are no taxi calls yet.",
    "noTowCalls": "There are no tow calls yet.",
    "postal": "Postal",
    "purge": "Purge",
    "purgeSelected": "Purge selected",
    "purgeSelectedCalls": "Purge selected calls",
    "removeMarker": "Remove marker",
    "selectCitizen": "Select citizen",
    "selectUnit": "Select Unit",
    "setMarker": "Set marker",
    "situationCode": "Situation Code",
    "taxi": "Taxi",
    "taxiCallCreated": "Successfully created the taxi call.",
    "toggleCall": "Toggle call",
    "tow": "Tow",
    "towCallCreated": "Successfully created the tow call.",
    "towLogs": "Tow Logs",
    "unassignFromCall": "Unassign from call",
    "viewEvents": "View Events",
  },
  "Common": {
    "accept": "Accept",
    "actions": "Actions",
    "cancel": "Cancel",
    "categories": "Categories",
    "citizen": "Citizen",
    "continue": "Continue",
    "create": "Create",
    "createdAt": "Created At",
    "crop": "Crop",
    "cropImage": "Crop Image",
    "decline": "Decline",
    "delete": "Delete",
    "description": "Description",
    "disabled": "Disabled",
    "edit": "Edit",
    "enabled": "Enabled",
    "error": "An unexpected error occurred",
    "events": "Events",
    "filters": "Filters",
    "goBack": "Go back",
    "hide": "Hide",
    "image": "Image",
    "imgur": "Imgur URL",
    "insufficientPermissions": "Insufficient Permissions",
    "manage": "Manage",
    "name": "Name",
    "no": "No",
    "noOptions": "No Options",
    "none": "None",
    "off": "Off",
    "on": "On",
    "openInSearch": "Open in search",
    "optionalField": "Optional",
    "postal": "Postal",
    "reason": "Reason",
    "reset": "Reset",
    "roleplayStopped": "Admins issued all players to stop all roleplay.",
    "save": "Save",
    "search": "Search",
    "show": "Show",
    "stopRoleplay": "Stop Roleplay",
    "success": "Success",
    "title": "Title",
    "type": "Type",
    "updatedAt": "Updated At",
    "view": "View",
    "viewDescription": "View description",
    "yes": "Yes",
  },
  "Errors": {
    "Error": "An unexpected error occurred. Please report this issue to an administrator or try again later.",
    "Invalid Permissions": "You don't have permissions to access this API Route.",
    "NOT_FOUND": "The requested item or page was not found. Please try refreshing the page or try again later.",
    "Network Error": "Could not reach the API due to a network error. Make sure that the API is online and that you are connected to the internet.",
    "allowRegularLoginIsDisabled": "Regular login is disabled. Please use Discord OAuth instead",
    "alreadyInThisBusiness": "You are already an employee at this business.",
    "alreadyPendingNameChange": "This citizen already has a pending name change request.",
    "businessCreatedButPending": "The business has been created successfully but is still awaiting approval to be used. It must be approved by an administrator.",
    "businessIsPending": "This business is still awaiting approval to be used. Try again later",
    "cannotRegisterFirstWithDiscord": "The first account must be registered without Discord. You can link Discord once the account is created.",
    "citizenNotFound": "That citizen was not found",
    "currentPasswordIncorrect": "Current password is incorrect",
    "dateLargerThanNow": "Please provide a date smaller than the current date.",
    "discordAccountAlreadyLinked": "This Discord account is already linked to another account.",
    "discordIdInUse": "This discord ID is already connected to a different account.",
    "discordNameInUse": "Could not create a new account with Discord since the username is already in-use.",
    "divisionNotInDepartment": "This division does not exist in this department.",
    "insufficientPermissions": "Insufficient Permissions",
    "invalidImageType": "Only types image/png, image/jpeg, image/jpg and image/gif are supported",
    "invalidRegistrationCode": "The registration code you provided is invalid.",
    "maxBusinessesLength": "You have reached the maximum limit of businesses per citizen.",
    "maxCitizensPerUserReached": "This user reached the maximum limit of citizens.",
    "maxDepartmentsReachedPerUser": "You've reached the max set of units with this department. Please choose a different department or try again later.",
    "maxDivisionsReached": "This officer reached the maximum limit of divisions per officer.",
    "maxLimitOfficersPerUserReached": "This user reached the maximum limit of officers.",
    "mustSetBotTokenGuildId": "This feature cannot be used because there was no \`BOT_TOKEN\` or \`GUILD_ID\` present. Please read how to enable Discord Roles: https://cad-docs.caspertheghost.me/docs/discord-integration/discord-roles",
    "nameAlreadyTaken": "A citizen with that full name is already in use.",
    "nameChangeRequestNotNew": "To create a name change request, the new name cannot match the previous name.",
    "notFound": "The requested item was not found.",
    "officerAlreadyMerged": "This unit is already marked as a combined unit.",
    "officerIsCombined": "This officer is marked as combined unit and cannot be used individually right now.",
    "passwordIncorrect": "The password you entered was not correct.",
    "passwordsDoNotMatch": "Passwords do not match",
    "plateAlreadyInUse": "That plate is already being used on a vehicle.",
    "recordOrWarrantAlreadyLinked": "A record or a warrant has already been requested. Please remove the already requested record/warrant.",
    "selectImageFirst": "You must first select an image.",
    "totpCodeRequired": "Please enter your 2FA code",
    "unitSuspended": "This unit is suspended. Please contact your supervisor.",
    "unknown": "An unexpected error occurred",
    "userAlreadyExists": "A user with that username already exists.",
    "userBanned": "This account is banned from this CAD.",
    "userDeleted": "This account has been deleted from this CAD.",
    "userNotFound": "That user was not found.",
    "whitelistDeclined": "This account was declined access to the CAD.",
    "whitelistPending": "This account is still pending access to the CAD.",
  },
  "Management": {
    "FEATURES": "Features",
    "GENERAL_SETTINGS": "General Settings",
    "IMPORT_CITIZENS": "Import Citizens",
    "IMPORT_VEHICLES": "Import Vehicles",
    "IMPORT_WEAPONS": "Import Weapons",
    "MANAGE_BUSINESSES": "Manage Businesses",
    "MANAGE_CAD_SETTINGS": "CAD Settings",
    "MANAGE_CITIZENS": "Manage Citizens",
    "MANAGE_EXPUNGEMENT_REQUESTS": "Manage Expungement Requests",
    "MANAGE_NAME_CHANGE_REQUESTS": "Manage Name Change Requests",
    "MANAGE_UNITS": "Manage Units",
    "MANAGE_USERS": "Manage Users",
    "MISC_SETTINGS": "Misc Settings",
    "action": "Action",
    "adminDashboard": "Admin Dashboard",
    "alert_deleteBusiness": "Are you sure you want to delete this <span>{name}</span>? All it's data will be lost. This action cannot be undone.",
    "allBusinesses": "All Businesses",
    "allUnits": "All Units",
    "allUsers": "All Users",
    "continue": "Yes, continue.",
    "declineOfficer": "Decline Officer",
    "deleteBusiness": "Delete Business",
    "departmentWhitelisting": "Department Whitelisting ({length})",
    "giveTempPassword": "Give Temporary Password",
    "import": "Import",
    "importViaFile": "Import via file",
    "inBolo": "In BOLO",
    "info_pendingBusinesses": "These are created businesses that are awaiting approval to be used. Accept or decline them.",
    "management": "Management",
    "noBusinesses": "There are no businesses for this CAD yet.",
    "noCitizens": "There are no citizens for this CAD yet.",
    "noPendingOfficers": "There are no officers pending access to a department.",
    "owner": "Owner",
    "pendingBusinesses": "Pending Businesses",
    "pendingUsers": "Pending Users",
    "reGenerateToken": "Re-generate Token",
    "status": "Status",
    "user": "User",
    "viewInfo": "View Info",
    "whitelisted": "Whitelisted",
  },
  "Nav": {
    "account": "Account",
    "admin": "Admin",
    "bleeter": "Bleeter",
    "business": "Business",
    "callHistory": "Call History",
    "citizen": "Citizen",
    "citizenLogs": "Citizen Logs",
    "citizens": "Citizens",
    "courthouse": "Courthouse",
    "dashboard": "Dashboard",
    "dispatch": "Dispatch",
    "emsFd": "EMS/FD",
    "impoundLot": "Impound Lot",
    "incidents": "Incidents",
    "jail": "Jail",
    "liveMap": "Live Map",
    "login": "Login",
    "logout": "Logout",
    "manageUnits": "Manage Units",
    "myDeputies": "My Deputies",
    "myOfficerLogs": "My Officer Logs",
    "myOfficers": "My Officers",
    "officer": "Officer",
    "register": "Register",
    "taxi": "Taxi",
    "tow": "Tow",
    "towLogs": "Tow Logs",
    "truckLogs": "Truck Logs",
  },
}
`;

exports[`Should return translations for citizen in RU 1`] = `
{
  "Citizen": {
    "address": "Адрес",
    "age": "Возраст",
    "alert_deleteCitizen": "Вы уверены, что хотите удалить <span>{citizen}</span>? Это действие необратимо.",
    "ccw": "Лицензия на скрытое ношение оружия",
    "citizen": "Гражданин",
    "citizens": "Граждане",
    "create911Call": "Вызвать 911",
    "createCitizen": "Создать гражданина",
    "createTaxiCall": "Вызвать такси",
    "createTowCall": "Вызвать эвакуатор",
    "dateOfBirth": "Дата рождения",
    "deleteCitizen": "Удалить",
    "driversLicense": "Водительская лицензия",
    "driversLicenseCategory": "Категории водительской лицензии",
    "editCitizen": "Редактировать",
    "ethnicity": "Этническая принадлежность",
    "eyeColor": "Цвет глаз",
    "fullName": "Полное имя",
    "gender": "Пол",
    "hairColor": "Цвет волос",
    "height": "Рост",
    "image": "Фото",
    "licenses": "Лицензии",
    "manageLicenses": "Управление лицензиями",
    "manageOccupation": "Редактировать род занятий",
    "name": "Имя",
    "occupation": "Род занятий",
    "phoneNumber": "Номер телефона",
    "pilotLicense": "Летная лицензия",
    "pilotLicenseCategory": "Категории летной лицензии",
    "registerVehicle": "Зарегистрировать транспорт",
    "registerWeapon": "Зарегистрировать оружие",
    "socialSecurityNumber": "Номер социального страхования",
    "surname": "Фамилия",
    "userNoCitizens": "У вас нет граждан",
    "viewCitizen": "Просмотр",
    "weaponLicense": "Лицензии на оружие",
    "weight": "Вес",
  },
  "Common": {
    "accept": "Принять",
    "actions": "Действия",
    "cancel": "Отменить",
    "categories": "Категории",
    "citizen": "Гражданин",
    "continue": "Продолжить",
    "create": "Создать",
    "createdAt": "Создан",
    "crop": "Обрезка",
    "cropImage": "Обрезать изображение",
    "decline": "Отклонить",
    "delete": "Удалить",
    "description": "Описание",
    "disabled": "Выключить",
    "edit": "Редактировать",
    "enabled": "Включить",
    "error": "Произошла непредвиденная ошибка",
    "events": "События",
    "hide": "Скрыть",
    "image": "Изображение",
    "imgur": "Ссылка на imgur",
    "insufficientPermissions": "Недостаточно прав",
    "manage": "Управлять",
    "name": "Название",
    "no": "Нет",
    "noOptions": "Нет вариантов",
    "none": "Не указан",
    "off": "Выкл",
    "on": "Вкл",
    "optionalField": "Опционально",
    "postal": "Почтовый код",
    "reason": "Причина",
    "reset": "Сброс",
    "roleplayStopped": "Администрация запретила отыгрыш РП.",
    "save": "Сохранить",
    "search": "Поиск",
    "show": "Показать",
    "stopRoleplay": "Остановить РП",
    "title": "Заголовок",
    "type": "Тип",
    "updatedAt": "Изменен",
    "view": "Просмотр",
    "viewDescription": "Просмотр описания",
    "yes": "Да",
  },
  "Errors": {
    "Error": "Произошла непредвиденная ошибка. Пожалуйста, сообщите об этой проблеме администратору или попробуйте позже.",
    "Invalid Permissions": "У вас нет прав для использования этого маршрута API.",
    "NOT_FOUND": "Запрошенная запись или страница не найдены. Пожалуйста, попробуйте обновить страницу или попробуйте позже.",
    "Network Error": "Не удалось получить доступ к API из-за сетевой ошибки. Убедитесь, что API находится в сети и что вы подключены к интернету.",
    "allowRegularLoginIsDisabled": "Обычный вход отключен. Пожалуйста, используйте Discord OAuth",
    "alreadyInThisBusiness": "Вы уже работаете в этой компании.",
    "businessCreatedButPending": "Компания создана, но ожидает подтверждения от администратора.",
    "businessIsPending": "Эта компания ожидает подтверждения. Попробуйте позже",
    "cannotRegisterFirstWithDiscord": "Первый аккаунт должен быть зарегистрирован без Discord. Вы можете привязать Discord после создания учетной записи.",
    "citizenNotFound": "Гражданин не найден",
    "currentPasswordIncorrect": "Неверный пароль",
    "dateLargerThanNow": "Пожалуйста, укажите дату меньше, чем сегодняшняя.",
    "discordAccountAlreadyLinked": "Этот Discord аккаунт уже привязан к другому пользователю.",
    "discordIdInUse": "Этот Discord ID уже подключен к другому пользователю.",
    "discordNameInUse": "Нельзя создать нового пользователя с таким Discord, поскольку имя уже используется.",
    "divisionNotInDepartment": "Такого подразделения в департаменте нет.",
    "insufficientPermissions": "Недостаточно прав",
    "invalidImageType": "Поддерживаются только типы image/png, image/jpeg, image/jpg и image/gif",
    "invalidRegistrationCode": "Неверный код регистрации.",
    "maxBusinessesLength": "Вы достигли максимального количества компаний для гражданина.",
    "maxCitizensPerUserReached": "Достигнут максимальный лимит по количеству граждан.",
    "maxDepartmentsReachedPerUser": "Достигнуто максимальное количество юнитов для департамента. Пожалуйста, выберите другой департамент или попробуйте позже.",
    "maxDivisionsReached": "Достигнут максимальный лимит по количеству подразделений для офицера.",
    "maxLimitOfficersPerUserReached": "Достигнут максимальный лимит по количеству офицеров.",
    "mustSetBotTokenGuildId": "Эта функция не может быть использована, так как \`BOT_TOKEN\` или \`GUILD_ID\` не указаны. Пожалуйста, прочтите, как включить поддержку ролей Discord: https://cad-docs.caspertheghost.me/docs/discord-integration/discord-roles",
    "nameAlreadyTaken": "Гражданин с таким полным именем уже существует.",
    "notFound": "Запись не найдена.",
    "officerAlreadyMerged": "Юнит уже отмечен, как парный.",
    "officerIsCombined": "Этот офицер используется в парном юните и не может быть использован отдельно сейчас.",
    "passwordIncorrect": "Введенный пароль неверен.",
    "passwordsDoNotMatch": "Пароли не совпадают",
    "plateAlreadyInUse": "Номерной знак уже используется.",
    "recordOrWarrantAlreadyLinked": "Запись/ордер уже запрошены. Пожалуйста, удалите запрошенную запись/ордер.",
    "selectImageFirst": "Сначала выберите изображение.",
    "totpCodeRequired": "Введите 2FA код",
    "unitSuspended": "Юнит заблокирован. Свяжитесь с супервайзером.",
    "unknown": "Произошла непредвиденная ошибка",
    "userAlreadyExists": "Пользователь с таким именем уже существует.",
    "userBanned": "Пользователь забанен в CAD.",
    "userDeleted": "Пользователь удален из CAD.",
    "userNotFound": "Пользователь не найден.",
    "whitelistDeclined": "Доступ к CAD отклонен.",
    "whitelistPending": "Пользователь ожидает подтверждения для доступа к CAD.",
  },
  "MedicalRecords": {
    "addMedicalRecord": "Добавить медицинскую запись",
    "alert_deleteMedicalRecord": "Вы уверены, что хотите удалить эту медицинскую запись? Это действие необратимо.",
    "bloodGroup": "Группа крови",
    "citizen": "Гражданин",
    "deleteMedicalRecord": "Удалить медицинскую запись",
    "diseases": "Заболевания",
    "editMedicalRecord": "Редактировать медицинскую запись",
    "noMedicalRecords": "У вас нет медицинских записей",
    "yourMedicalRecords": "Медицинские записи",
  },
  "Nav": {
    "account": "Аккаунт",
    "admin": "Администратор",
    "bleeter": "Bleeter",
    "business": "Компания",
    "callHistory": "История вызовов",
    "citizen": "Гражданин",
    "citizenLogs": "Журнал граждан",
    "citizens": "Граждане",
    "courthouse": "Суд",
    "dashboard": "Панель управления",
    "dispatch": "Диспетчер",
    "emsFd": "EMS/FD",
    "impoundLot": "Штрафстоянки",
    "incidents": "Инциденты",
    "jail": "Тюрьма",
    "liveMap": "Live Map",
    "login": "Вход",
    "logout": "Выход",
    "manageUnits": "Управление юнитами",
    "myDeputies": "Сотрудники",
    "myOfficerLogs": "Журнал офицеров",
    "myOfficers": "Офицеры",
    "officer": "Офицер",
    "register": "Регистрация",
    "taxi": "Такси",
    "tow": "Эвакуатор",
    "towLogs": "Журнал эвакуации",
    "truckLogs": "Журнал грузоперевозок",
  },
  "Vehicles": {
    "addVehicle": "Добавить транспорт",
    "alert_deleteVehicle": "Вы уверены, что хотите удалить транспорт? Это действие необратимо.",
    "business": "Компания",
    "color": "Цвет",
    "deleteVehicle": "Удалить транспорт",
    "insuranceStatus": "Статус страховки",
    "model": "Модель",
    "noVehicles": "У вас нет транспорта.",
    "owner": "Владелец",
    "plate": "Номерной знак",
    "registeredVehicles": "Зарегистрированный транспорт",
    "registrationStatus": "Статус регистрации",
    "reportAsStolen": "Отметить, как угнанный",
    "vinNumber": "VIN",
    "yourVehicles": "Ваш транспорт",
  },
  "Weapons": {
    "addWeapon": "Добавить оружие",
    "alert_deleteWeapon": "Вы уверены, что хотите удалить это оружие? Это действие необратимо.",
    "deleteWeapon": "Удалить оружие",
    "model": "Модель",
    "noWeapons": "У вас нет зарегистрированного оружия.",
    "registeredWeapons": "Зарегистрированное оружие",
    "registrationStatus": "Статус регистрации",
    "serialNumber": "Серийный номер",
    "yourWeapons": "Ваше оружие",
  },
}
`;
